compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3839,6 +3850,18 @@ object Types {
         superType
     }
 
+    inline def map(inline op: Type => Type)(using Context) =
+      def mapArgs(args: List[Type]): List[Type] = args match
+        case args @ (arg :: rest) => args.derivedCons(op(arg), mapArgs(rest))
+        case nil => nil
+      derivedAppliedType(op(tycon), mapArgs(args))
+
+    inline def fold[T](x: T, inline op: (T, Type) => T)(using Context): T =
+      def foldArgs(x: T, args: List[Type]): T = args match
+        case arg :: rest => foldArgs(op(x, arg), rest)
+        case nil => x
+      foldArgs(op(x, tycon), args)
```

Comment:
In `map` and `fold`, do we need to recurse on `typcon`? The semantics is quite different from `TypeMap` and `TypeAccumulator`, and is specific to `Substituters`. Maybe add a note here to avoid misuse?